docs/proposed/http-storage-node-protocol.rst

```diff
@@ -0,0 +1,506 @@
+
+Storage nodes will begin to operate a new GBS server.
+They may re-use their existing x509 certificate or generate a new one.
+Generation of a new certificate allows for certain non-optimal conditions to be addressed:
+
+* The ``commonName`` of ``newpb_thingy`` may be changed to a more descriptive value.
+* A ``notValidAfter`` field with a timestamp in the past may be updated.
+
+Storage nodes will announce a new fURL for this new HTTP-based server.
+This fURL will be announced alongside their existing Foolscap-based server's fURL.
+Such an announcement will resemble this::
+
+  {
+      "anonymous-storage-FURL": "pb://...",          # The old key
+      "anonymous-storage-gbs-FURL": "pb://...#v=2"   # The new key
+  }
+
+The transition process will proceed in three stages:
+
+1. The first stage represents the starting conditions in which clients and servers can speak only Foolscap.
+#. The intermediate stage represents a condition in which some clients and servers can both speak Foolscap and GBS.
+#. The final stage represents the desired condition in which all clients and servers speak only GBS.
+
+During the first stage only one client/server interaction is possible:
+the storage server announces only Foolscap and speaks only Foolscap.
+During the final stage there is only one supported interaction:
+the client and server are both updated and speak GBS to each other.
+
+During the intermediate stage there are four supported interactions:
+
+1. Both the client and server are non-updated.
+   The interaction is just as it would be during the first stage.
+#. The client is updated and the server is non-updated.
+   The client will see the Foolscap announcement and the lack of a GBS announcement.
+   It will speak to the server using Foolscap.
+#. The client is non-updated and the server is updated.
+   The client will see the Foolscap announcement.
+   It will speak Foolscap to the storage server.
+#. Both the client and server are updated.
+   The client will see the GBS announcement and disregard the Foolscap announcement.
+   It will speak GBS to the server.
+
+There is one further complication:
+the client maintains a cache of storage server information
+(to avoid continuing to rely on the introducer after it has been introduced).
+The follow sequence of events is likely:
+
+1. The client connects to an introducer.
+#. It receives an announcement for a non-updated storage server (Foolscap only).
+#. It caches this announcement.
+#. At some point, the storage server is updated.
+#. The client uses the information in its cache to open a Foolscap connection to the storage server.
+
+Ideally,
+the client would not rely on an update from the introducer to give it the GBS fURL for the updated storage server.
+Therefore,
+when an updated client connects to a storage server using Foolscap,
+it should request the server's version information.
+If this information indicates that GBS is supported then the client should cache this GBS information.
```

Comment:
Does the "version information" allow us to do feature-flag style things here? e.g. conceptually like ``features: {"great-black-swamp": 1}`` or so..? (I greatly prefer that to comparing actual versions, I think?)